Tutorials/Creating-an-XML-Site-Map/index.md

```diff
@@ -0,0 +1,336 @@
+```
+
+we have a full sitemap rendered for the site!
+
+![Example of Sitemap with children](images/sitemapWithChildren.png)
+
+#### Checking if a page should be on the sitemap
+
+This is all very well, but what if some super secret pages shouldn't be on the sitemap, what about the document type blacklist we mentioned earlier, what if we only want to go 3 levels deep?
+
+##### HideFromSiteMap
+
+We added a hideFromXmlSitemap checkbox to hopefully all of our document types, let's update the helper to only return children that haven't got the checkbox set, excluding these pages (and any beneath them) from the sitemap.
+
+```csharp
+@helper RenderSiteMapUrlEntriesForChildren(IPublishedContent parentPage)
+{          
+    foreach (var page in parentPage.Children.Where(f=>!f.GetPropertyValue<bool>("hideFromXmlSiteMap")))
+    {
+        @RenderSiteMapUrlEntry(page)
+        if (page.Children.Any(f=>!f.GetPropertyValue<bool>("hideFromXmlSiteMap"))){
+            @RenderSiteMapUrlEntriesForChildren(page)
+        }               
+    }
+}
+```
+
+##### Depth
+
+What if we only want the sitemap to go three levels deep?
+
+We added a property on the XmlSiteMap document type called depth... we can read that in our template and pass it to our helper:
+
+```csharp
+@inherits Umbraco.Web.Mvc.UmbracoTemplatePage
+@{Layout = null; Response.ContentType = "text/xml"; int maxSiteMapDepth = Model.Content.HasValue("maxSiteMapDepth") ? Model.Content.GetPropertyValue<int>("maxSiteMapDepth") : int.MaxValue; IPublishedContent siteHomePage = Model.Content.Site(); }<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemalocation="http://www.google.com/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">@RenderSiteMapUrlEntry(siteHomePage)@RenderSiteMapUrlEntriesForChildren(siteHomePage, maxSiteMapDepth)</urlset>
+
+.....
+
+@helper RenderSiteMapUrlEntriesForChildren(IPublishedContent parentPage, int maxSiteMapDepth)
+{          
+    foreach (var page in parentPage.Children.Where(f=>!f.GetPropertyValue<bool>("hideFromXmlSiteMap"))
+    {
+        @RenderSiteMapUrlEntry(page)
+        if (page.Level < maxSiteMapDepth && page.Children.Any(f=>!f.GetPropertyValue<bool>("hideFromXmlSiteMap")){
+            @RenderSiteMapUrlEntriesForChildren(page, maxSiteMapDepth)
+        }               
+    }
+}
+```
+
+##### DocumentType Blacklist
+
+Automatically exclude certain document types regardless of their 'hide from sitemap' setting.
+
+```csharp
+string blacklistedDocumentTypeList = Model.Content.GetPropertyValue<string>("documentTypeBlackList");
+string[] blackListedDocumentTypes = (!String.IsNullOrEmpty(blacklistedDocumentTypeList)) ? blacklistedDocumentTypeList.split(new char[]{','}, StringSplitOptions.RemoveEmptyEntries) : new string[];
```

Comment:
On this step I get an error on "new string[];" That I can't quite figure out how to solve:

Compiler Error Message: CS1586: Array creation must have array size or array initializer

Do you have any ideas how we could rewrite this to prevent that error? 😁

Reply:
@jmayntzhusen just missing the curly brackets :-)

eg new string[] {};

will do it....